### DISCUSSION TITLE
think about a solution to add your own containers.json files

### DISCUSSION THREAD

**Discussion by szaimen** [nextcloud/all-in-one#1581] [Ideas] [answered]
Required:
- [x] add docs https://github.com/nextcloud/all-in-one/pull/3431
- [x] additionally should additoonal containers alway rely on the apache container so that they get started at all (if display_name not empty, add to the list) https://github.com/nextcloud/all-in-one/pull/3425
- [x] merge services array with https://www.php.net/manual/en/function.array-merge-recursive.php? https://github.com/nextcloud/all-in-one/pull/3425
- [x] allow to enable apps by specifying its name via variable, then choose file automatically https://github.com/nextcloud/all-in-one/pull/3424
- [x] community containers must be bundled in AIO during build time. By doing so we make sure that containers.json is always valid https://github.com/nextcloud/all-in-one/pull/3424
- [x] Add minlenght to json schema strings: https://stackoverflow.com/a/53733250 https://github.com/nextcloud/all-in-one/pull/2429
- [x] Also add other types or regex? https://json-schema.org/understanding-json-schema/index.html# https://github.com/nextcloud/all-in-one/pull/2429
- [x] remove https://github.com/nextcloud/all-in-one/blob/3e08843b1474c49e651bd6344e48e33ed8e270e5/php/composer.json#L20 again since we test with CI and bundle things https://github.com/nextcloud/all-in-one/pull/3417
- [ ] as a follow-up, add some documentation on this: e.g. where containers need to get deployed (on docker hub), which are tags that are needed (maybe also link to some giithub actions), what is currently possible regarding env variables, etc. in best case should you run docker rootless and obviously only add definitions of trusted people
- [x] ~~merge expose and internal_port of containeers definition or rename internal_port to network_mode~~ not needed since expose is not needed for AIO to work. Only an internal thing for kubernetes to work. So still only internal_port is used for checking the availability
- [x] maybe even allow to run docker exec commands that get executed after all containers are running. (Disable setting on-click instance settings) https://github.com/nextcloud/all-in-one/pull/2454
- [x] for reverse proxy we need to provide a way to expand the conf. Could probably use the caddy import statement for this: https://caddyserver.com/docs/caddyfile/directives/import. For the template itself should be added to an additional volume that could be mounted into the apache container and mounted by the additional container with the restriction that only subdirectories are supported. For anything that needs a domain should probably rather a reverse proxy get used  https://github.com/nextcloud/all-in-one/pull/2457
- [ ] add some docs for the feature above
- [ ] add docs for host mode container feature
- [x] We could add a way to allow to run a reverse proxy from the custom containers.json (hostmode would be needed then; this would allow much more customization) (by setting internal_port to "host"
- [ ] add documentation on global variables and how to use them.
- [x] allow to add devices to container.json
- [x] allow to define CapAdd in containers.json https://github.com/nextcloud/all-in-one/pull/2113
- [x] allow to enable securityopt: apparmor=undefined https://github.com/nextcloud/all-in-one/pull/2432
- [x] validate json before merging it with default containers.json: https://github.com/justinrainbow/json-schema https://github.com/nextcloud/all-in-one/pull/2611
- [x] allow to define if a volume gets included in backup and restore e.g. via include_in_backup_and_restore https://github.com/nextcloud/all-in-one/pull/2455
- [x] https://github.com/nextcloud/all-in-one/discussions/490#discussioncomment-4720018
- [x] add profiles and networks to containers definition so that it can be removed form update-yaml.sh https://github.com/nextcloud/all-in-one/pull/2631
- [x] add image_tag to definition-schema in order to allow to pin the container tag and not rely on the chosen channel. Also needs to be used for starting container and update check, etc. https://github.com/nextcloud/all-in-one/pull/3191
- [x] also use image_tag in containers.json for better testing? Would need a new variable then so that it gets subsituted automatically. In that case image_tag should be required https://github.com/nextcloud/all-in-one/pull/3314

Follow-ups:
- [ ] think about instance restore: should teh backup container run twice in a row to restore the setting first and then start the correct restore?
- [ ] Only add 
	- [ ] vaultwarden
	- [ ] plex
	- [ ] smbserver
	- [ ] caddy with geoblocking
	- [ ] disk monitoring? (better do from host)
	- [ ] pihole
	- [ ] pivpn (use fritzbox)
	- [ ] remotedesktop
	- [x] fail2ban https://github.com/nextcloud/all-in-one/pull/2637
- [ ] instead of modifying container to put a caddy import into place, should there be an option to modify a subpath that is automatically forwarded to the correct container in a subpath available on the nextcloud domain? https://github.com/nextcloud/all-in-one/pull/3192 (maybe also use that in the containers but must also work without)
- [ ] allow to define if a secret gets shown in the aio interface and also allow to add a description for a secret and also allow to add a description of what the token does
- [ ] allow to set not defined environmental variables from the aio interface: by providing an input field and regex and maybe even description).
- [ ] allow to define custom cronjobs

PRs:
- https://github.com/nextcloud/all-in-one/pull/1595
- https://github.com/nextcloud/all-in-one/pull/1597
- https://github.com/nextcloud/all-in-one/pull/1598
- https://github.com/nextcloud/all-in-one/pull/1599
- https://github.com/nextcloud/all-in-one/pull/1603
- https://github.com/nextcloud/all-in-one/pull/1604
- https://github.com/nextcloud/all-in-one/pull/1635
- https://github.com/nextcloud/all-in-one/pull/1637
- https://github.com/nextcloud/all-in-one/pull/1659
- https://github.com/nextcloud/all-in-one/pull/2029
- https://github.com/nextcloud/all-in-one/pull/2113
- https://github.com/nextcloud/all-in-one/pull/2429
- https://github.com/nextcloud/all-in-one/pull/2432
- https://github.com/nextcloud/all-in-one/pull/2637

**Comment by szaimen** [comment]
Btw @Zoey2936 would this feature also be something that you would be interested in? :)

**Reply by Zoey2936** [reply]
what do you mean exactly?

**Reply by szaimen** [reply]
I mean being able to add additional containers to AIO :)

**Reply by Zoey2936** [reply]
From my side I have no use case for this, so I would say that this is a nice feature, but I don't need it.

**Reply by szaimen** [reply]
I see 👍

**Comment by szaimen** [comment, marked as answer]
Required things are done. Follow-ups in https://github.com/nextcloud/all-in-one/discussions/3432

**Reply by szaimen** [reply]
This is now released with v7.4.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel
